sdk/core/core/src/az_client_secret_credential.c

```diff
@@ -43,125 +43,75 @@ static AZ_NODISCARD az_result az_token_credential_get_token(
+      az_span_builder builder
+          = az_span_builder_create((az_mut_span)AZ_SPAN_FROM_ARRAY(auth_url_buf));
 
-  AZ_CONTRACT(auth_url_maxsize <= (size_t) ~(uint16_t)0, AZ_ERROR_ARG);
+      AZ_RETURN_IF_FAILED(az_span_builder_append(&builder, auth_url1));
+      AZ_RETURN_IF_FAILED(az_uri_encode(credential->tenant_id, &builder));
+      AZ_RETURN_IF_FAILED(az_span_builder_append(&builder, auth_url2));
 
-  size_t const headers_size = sizeof(az_pair) + 7; // We need 7 because our code aligns at 8 byte boundary
-  {
-    AZ_CONTRACT(
-        response_buf.size >= AZ_TOKEN_CREDENTIAL_GET_TOKEN_MIN_BUFFER, AZ_ERROR_BUFFER_OVERFLOW);
-
-    size_t const request_elements[] = {
-      credential->tenant_id.size * AZ_TOKEN_CREDENTIAL_GET_TOKEN_URLENCODE_FACTOR,
-      credential->client_id.size * AZ_TOKEN_CREDENTIAL_GET_TOKEN_URLENCODE_FACTOR,
-      credential->client_secret.size * AZ_TOKEN_CREDENTIAL_GET_TOKEN_URLENCODE_FACTOR,
-      resource_url.size * AZ_TOKEN_CREDENTIAL_GET_TOKEN_URLENCODE_FACTOR,
-    };
-
-    size_t required_request_size = auth_url1.size + auth_url2.size + auth_body1.size
-        + auth_body2.size + auth_body3.size + headers_size;
-
-    for (size_t i = 0; i < AZ_ARRAY_SIZE(request_elements); ++i) {
-      required_request_size += request_elements[i];
-      AZ_CONTRACT(required_request_size > request_elements[i], AZ_ERROR_BUFFER_OVERFLOW);
+      auth_url = az_span_builder_result(&builder);
     }
 
-    AZ_CONTRACT(response_buf.size >= required_request_size, AZ_ERROR_BUFFER_OVERFLOW);
-  }
+    uint8_t auth_body_buf[200] = { 0 };
+    az_span auth_body = { 0 };
 
-  {
-    AZ_CONTRACT(!az_span_is_overlap(az_mut_span_to_span(response_buf), resource_url), AZ_ERROR_ARG);
+    static az_span const auth_body1 = AZ_CONST_STR("grant_type=client_credentials&client_id=");
+    static az_span const auth_body2 = AZ_CONST_STR("&client_secret=");
+    static az_span const auth_body3 = AZ_CONST_STR("&resource=");
 
-    AZ_CONTRACT(
-        !az_span_is_overlap(az_mut_span_to_span(response_buf), credential->tenant_id),
-        AZ_ERROR_ARG);
-
-    AZ_CONTRACT(
-        !az_span_is_overlap(az_mut_span_to_span(response_buf), credential->client_id),
-        AZ_ERROR_ARG);
-
-    AZ_CONTRACT(
-        !az_span_is_overlap(az_mut_span_to_span(response_buf), credential->client_secret),
-        AZ_ERROR_ARG);
-  }
-
-  az_span auth_url = { 0 };
-  az_span auth_body = { 0 };
-  {
-    az_span_builder builder = az_span_builder_create(response_buf);
-
-    AZ_RETURN_IF_FAILED(az_span_builder_append(&builder, auth_url1));
-    AZ_RETURN_IF_FAILED(az_uri_encode(credential->tenant_id, &builder));
-    AZ_RETURN_IF_FAILED(az_span_builder_append(&builder, auth_url2));
-
-    auth_url = az_span_builder_result(&builder);
-    builder = az_span_builder_create(az_mut_span_drop(response_buf, auth_url.size + headers_size));
+    az_span_builder builder
+        = az_span_builder_create((az_mut_span)AZ_SPAN_FROM_ARRAY(auth_body_buf));
```

Comment:
Rather than casting the span to a az_mut_span, add AZ_MUT_SPAN_FROM_ARRAY

Reply:
`AZ_SPAN_FROM_ARRAY` doesn't have a type. So, currently we always need to specify a type whatever it's `az_span`, `az_mut_span` or `az_pair_span`. I see two problems. 
1. The name `az_span` is not good, because we have different spans, I would recommend to rename it to `az_byte_span`
2. redesign the `AZ_SPAN_FROM_ARRAY`. May be it should take a type and an array  `AZ_SPAN_FROM_ARRAY(TYPE, ARRAY)`. For example `az_span_builder_create(AZ_SPAN_FROM_ARRAY(az_mut_span, auth_body_buf));

Reply:
Agree wit hSergey, but now the code is rewritten in a way it is even no longer needed.